File: ui/lib/output/analysis_series_export.dart
import 'package:excel/excel.dart';
import 'package:geneweb/analysis/analysis_series.dart';

/// Responsible for exporting the [series]
class AnalysisSeriesExport {
  final AnalysisSeries series;

  AnalysisSeriesExport(this.series);

  /// Exports the series to Excel
  Future<List<int>?> toExcel(String fileName, Function(double progress) progressCallback) async {
    assert(series.geneList.genes.isNotEmpty);
    var excel = Excel.createExcel();
    final originalSheets = excel.sheets.keys;
    final headerCellStyle = CellStyle(backgroundColorHex: ExcelColor.fromHexString('FFDDFFDD'), bold: true);

    // selected_genes sheet
    Sheet genesSheet = excel['selected_genes'];
    final stages = series.geneList.genes.first.transcriptionRates.keys.toList();
    // header row
    genesSheet.appendRow([
      const TextCellValue('Gene Id'),
      const TextCellValue('Matches'),
      for (final stage in stages) TextCellValue(stage),
    ]);
    // data rows
    final resultsMap = series.resultsMap;
    for (var i = 0; i < series.geneList.genes.length; i++) {
      if (i % 1000 == 0) {
        progressCallback(i / series.geneList.genes.length * 0.5);
        await Future.delayed(const Duration(milliseconds: 20));
      }
      final gene = series.geneList.genes[i];
      genesSheet.appendRow([
        TextCellValue(gene.geneId),
        IntCellValue(resultsMap[gene.geneId]?.length ?? 0),
//        series.result?.where((g) => g.gene.geneId == gene.geneId).length,
        for (final stage in stages)
          gene.transcriptionRates[stage] == null
              ? const TextCellValue('')
              : DoubleCellValue(gene.transcriptionRates[stage]!.toDouble()),
      ]);
    }
    // style the header and first two columns
    for (int i = 0; i < genesSheet.maxColumns; i++) {
      genesSheet.cell(CellIndex.indexByColumnRow(columnIndex: i, rowIndex: 0)).cellStyle = headerCellStyle;
    }
    for (int i = 0; i < genesSheet.maxRows; i++) {
      if (i % 1000 == 0) {
        progressCallback(0.5 + i / genesSheet.maxRows * 0.1);
        await Future.delayed(const Duration(milliseconds: 20));
      }
      genesSheet.cell(CellIndex.indexByColumnRow(columnIndex: 0, rowIndex: i)).cellStyle = headerCellStyle;
      genesSheet.cell(CellIndex.indexByColumnRow(columnIndex: 1, rowIndex: i)).cellStyle = headerCellStyle;
    }

    // distribution sheet
    Sheet distributionSheet = excel['distribution'];
    distributionSheet.appendRow([const TextCellValue('Interval'), const TextCellValue('Genes with motif')]);
    int i = 0;
    final datapoints = series.distribution!.dataPoints!;
    for (final dataPoint in datapoints) {
      if (i++ % 100 == 0) {
        progressCallback(0.6 + i / datapoints.length * 0.3);
        await Future.delayed(const Duration(milliseconds: 20));
      }
      distributionSheet.appendRow([
        TextCellValue(dataPoint.label),
        for (final gene in dataPoint.genes) TextCellValue(gene),
      ]);
    }
    for (int i = 0; i < distributionSheet.maxColumns; i++) {
      distributionSheet.cell(CellIndex.indexByColumnRow(columnIndex: i, rowIndex: 0)).cellStyle = headerCellStyle;
    }
    for (int i = 0; i < distributionSheet.maxRows; i++) {
      if (i % 100 == 0) {
        progressCallback(0.9 + i / distributionSheet.maxRows * 0.1);
        await Future.delayed(const Duration(milliseconds: 20));
      }
      distributionSheet.cell(CellIndex.indexByColumnRow(columnIndex: 0, rowIndex: i)).cellStyle = headerCellStyle;
    }

    for (var element in originalSheets) {
      excel.delete(element);
    }
    return excel.save(fileName: fileName);
  }
}

File: backend/lib/output/analysis_series_export.py
import io
import asyncio
import xlsxwriter

class AnalysisSeriesExport:
    """
    Responsible for exporting a single AnalysisSeries to Excel.
    """

    def __init__(self, series: "AnalysisSeries"):
        """
        :param series: The AnalysisSeries to export
        """
        self.series = series

    async def to_excel(
            self,
            file_name: str,
            progress_callback
    ) -> bytes:
        """
        Exports the series to Excel in memory.
        Returns the raw bytes of the generated Excel file.
        :param file_name: The desired file name (not strictly needed for in-memory generation).
        :param progress_callback: a function(progress: float) to be called with progress [0..1].
        """

        # Create a BytesIO stream for in-memory Excel
        output = io.BytesIO()

        # Create a Workbook
        workbook = xlsxwriter.Workbook(output, {"in_memory": True})

        # Define a header cell format
        header_format = workbook.add_format({
            "bold": True,
            "bg_color": "#DDFFDD"
        })

        # 1) selected_genes sheet
        genes_sheet = workbook.add_worksheet("selected_genes")

        # Gather stage names from the first gene’s transcriptionRates
        genes = self.series.geneList.genes
        if not genes:
            # If somehow no genes exist, just return empty file
            workbook.close()
            return output.getvalue()
        stages = list(genes[0].transcriptionRates.keys())

        # -- Write header row
        header_row = ["Gene Id", "Matches"] + stages
        for col_idx, header_text in enumerate(header_row):
            genes_sheet.write(0, col_idx, header_text, header_format)

        # Build results map
        results_map = self.series.resultsMap

        # Write data rows
        for i, gene in enumerate(genes):
            # Progress ~ 0..0.5
            if i % 1000 == 0:
                progress = (i / len(genes)) * 0.5
                progress_callback(progress)
                await asyncio.sleep(0.02)  # small delay to mimic Dart's Future.delayed

            row_idx = i + 1
            genes_sheet.write(row_idx, 0, gene.geneId, header_format)  # style first col
            count_matches = len(results_map.get(gene.geneId, []))
            genes_sheet.write(row_idx, 1, count_matches, header_format)  # style second col

            # Write transcription rates
            for c, stage in enumerate(stages, start=2):
                val = gene.transcriptionRates.get(stage)
                if val is not None:
                    genes_sheet.write(row_idx, c, float(val))
                else:
                    genes_sheet.write(row_idx, c, "")  # empty cell

        # 2) distribution sheet
        distribution_sheet = workbook.add_worksheet("distribution")
        distribution_sheet.write(0, 0, "Interval", header_format)
        distribution_sheet.write(0, 1, "Genes with motif", header_format)

        data_points = []
        if self.series.distribution and self.series.distribution.dataPoints:
            data_points = self.series.distribution.dataPoints
        else:
            data_points = []

        # Write distribution data
        for i, dp in enumerate(data_points):
            # Progress ~ 0.6..0.9
            if i % 100 == 0:
                partial_progress = 0.6 + (i / max(len(data_points), 1)) * 0.3
                progress_callback(partial_progress)
                await asyncio.sleep(0.02)

            row_idx = i + 1
            # First cell is label, style it
            distribution_sheet.write(row_idx, 0, dp.label, header_format)

            # Then each gene in dp.genes. Notice in Dart code, we do:
            # [TextCellValue(dataPoint.label), for gene in dataPoint.genes => cell...]
            # That places each gene in a new column.
            # So let's do that here. The second col is the first gene, etc.
            col_idx = 1
            for gene_id in dp.genes:
                distribution_sheet.write(row_idx, col_idx, gene_id)
                col_idx += 1

        # The Dart code also modifies the default sheet(s), removing them at the end.
        # XlsxWriter by default doesn't create an extra "Sheet1" if we name our own sheets.
        # So we don't need to "delete" the original sheets.

        # Finish workbook
        workbook.close()
        output.seek(0)
        return output.getvalue()
